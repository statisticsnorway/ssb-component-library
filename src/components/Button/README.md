Button
========

> Action buttons

### Usage

#### HTML

```html
<button class="ssb-button">Secondary button</button>

<button class="ssb-btn primary-btn">Primary button</button>

<button class="ssb-btn primary-btn">
    <div class="sb-icon">
        <i>{feather.chevronDown 18px}</i>
    </div>Primary icon
</button>

<button class="ssb-btn negative">Secondary negative button</button>

<button disabled="true">Disabled button</button>
```

#### React

```jsx harmony
<Button>Secondary button</Button>

<Button icon={<ChevronDown />} primary>Primary icon</Button>

<Button primary>Primary button</Button>

<Button negative primary>Primary negative button</Button>

<Button disabled>Disabled button</Button>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| children | Required node | Button text or/and icon |
| className | String | Optional container class |
| disabled | bool | Decides if the button is disabled |
| icon | node | Renders an icon |
| negative | bool | Changes design |
| onClick | function | Callback for button click |
| primary | bool | Changes style to represent a primary button |
| type | string | Button type. Can be 'submit', 'reset', or 'button'. Defaults to 'button' |
