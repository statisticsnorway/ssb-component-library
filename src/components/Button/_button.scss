.ssb-btn {
  @include focus-marker;
  @include roboto;
  @include reset-button;
  align-items: center;
  background: $ssb-white;
  border: 2px solid $ssb-green-4;
  border-radius: 2px;
  color: $ssb-green-4;
  display: flex;
  font-size: 16px;
  font-weight: bold;
  height: fit-content;
  line-height: 1.25;
  min-height: 40px;
  padding: 12px 20px;
  text-align: center;
  text-underline-position: under;
  transition: background .18s, color .18s;
  white-space: nowrap;

  &:hover, &:active {
    cursor: pointer;
  }

  &:hover, &:focus {
    background: $ssb-green-4;
    color: $ssb-white;
    text-decoration: underline;
  }

  &:active {
    background: $ssb-dark-5;
    border: 2px solid $ssb-dark-5;
    color: $ssb-white;
  }

  &:disabled {
    background: $ssb-white;
    border-color: $ssb-dark-2;
    color: $ssb-dark-2;
    cursor: not-allowed;

    &:hover, &:focus {
      text-decoration: none;
    }
  }

  &.negative {
    background: $ssb-dark-5;
    color: $ssb-green-2;
    border-color: $ssb-green-2;

    &:hover, &:focus {
      background: $ssb-green-2;
      color: $ssb-dark-5;
    }

    &:active {
      background: $ssb-white;
      color: $ssb-dark-5;
    }

    &:disabled {
      background: $ssb-dark-5;
      border-color: $ssb-dark-4;
      color: $ssb-dark-4;

      &:hover, &:focus {
        text-decoration: none;
      }
    }
  }

  &.primary-btn {
    background: $ssb-green-4;
    border: 2px solid $ssb-green-4;
    color: $ssb-white;

    &:hover, &:active, &:focus {
      background: $ssb-dark-5;
      border: 2px solid $ssb-dark-5;
      text-decoration: underline;
    }

    &:disabled {
      background: $ssb-dark-2;
      border: 2px solid $ssb-dark-2;
      color: $ssb-white;
      cursor: not-allowed;
      font-weight: normal;
    }

    &.negative {
      background: $ssb-green-2;
      border: 2px solid $ssb-green-2;
      color: $ssb-dark-5;

      &:hover, &:active, &:focus  {
        background: $ssb-white;
        border: 2px solid $ssb-white;
      }

      &:disabled {
        background: $ssb-dark-4;
        border-color: $ssb-dark-4;
        color: $ssb-dark-5;
      }
    }
  }

  .sb-icon {
    display: inline-flex;
    font-size: 18px;
    margin-right: 5px;
  }
}
