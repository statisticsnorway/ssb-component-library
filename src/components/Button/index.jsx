import React from 'react';
import PropTypes from 'prop-types';

const Button = ({
	children, disabled, onClick, primary, type,
}) => (
	<button
		type={type}
		className={`ssb-btn ${primary ? 'primary-btn' : 'secondary-btn'}`}
		onClick={onClick}
		disabled={disabled}
	> {children}
	</button>
);

Button.defaultProps = {
	disabled: false,
	onClick: () => {},
	primary: false,
	type: 'button',
};

Button.propTypes = {
	children: PropTypes.node.isRequired,
	disabled: PropTypes.bool,
	onClick: PropTypes.func,
	primary: PropTypes.bool,
	type: PropTypes.string,
};

export default Button;
