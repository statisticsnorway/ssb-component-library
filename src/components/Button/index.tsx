import React, { forwardRef, type ReactNode } from 'react'

interface ButtonProps {
  children: ReactNode
  className?: string
  disabled?: boolean
  icon?: ReactNode
  negative?: boolean
  onClick?: () => void
  primary?: boolean
  type?: 'button' | 'submit' | 'reset'
  ariaLabel?: string
  onKeyDown?: () => void
}

const Button = forwardRef<HTMLButtonElement, ButtonProps>(
  (
    {
      children,
      className,
      disabled = false,
      icon,
      negative = false,
      onClick = () => {},
      primary = false,
      type = 'button',
      ariaLabel,
      onKeyDown = () => {},
    },
    ref
  ) => (
    <button
      type={type}
      ref={ref}
      className={`ssb-btn${negative ? ' negative' : ''} ${primary ? 'primary-btn' : 'secondary-btn'}${className ? ` ${className}` : ''}`}
      onClick={onClick}
      onKeyDown={onKeyDown}
      disabled={disabled}
      aria-label={ariaLabel || undefined}
    >
      {icon && <div className='sb-icon'>{icon}</div>}
      {children}
    </button>
  )
)

Button.displayName = 'Button'

export default Button
