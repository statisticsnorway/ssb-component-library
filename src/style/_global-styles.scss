body {
  @include open-sans;
  color: $ssb-dark-6;
}

.flex-row {
  display: flex;
  flex-direction: row;
}

.flex-column {
  display: flex;
  flex-direction: column;
}
