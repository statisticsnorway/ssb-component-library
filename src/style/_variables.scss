$ssb-green-1: #e5fee6;
$ssb-green-2: #b6e8b8;
$ssb-green-3: #1a9d49;
$ssb-green-4: #00824d;
$ssb-green-5: #075745;
$ssb-blue-3: #3396d2;
$ssb-red-4: #c4351c;
$negative-red: #f8a67d;
$ssb-dark-2: #a2baba;
$ssb-dark-3: #6f9090;
$ssb-dark-4: #4b7272;
$ssb-dark-5: #274247;
$ssb-dark-6: #162327;
$ssb-white: #fff;
$ice-cold-grey: #ecf8f9;
$ssb-purple-3: #9575ff;


$mobile: 'screen and (max-width: 768px)';
$tablet: 'screen and (max-width: 992px)';
$desktop: 'screen and (min-width: 992px)';

// Use media queries like this:
//@media #{$mobile} { @content }

$input-field-height: 36px;
$global-padding-size: 10px;
$max-content-size: 1200px;

@mixin roboto {
  color: $ssb-dark-6;
  font-family: 'Roboto', sans-serif;
  font-stretch: normal;
  font-weight: normal;
}

@mixin roboto-condenced {
  color: $ssb-dark-6;
  font-family: 'Roboto Condensed', sans-serif;
  font-stretch: condensed;
  font-weight: bold;
}

@mixin open-sans {
  color: $ssb-dark-6;
  font-family: 'Open Sans', sans-serif;
  font-stretch: normal;
  font-weight: normal;
}
