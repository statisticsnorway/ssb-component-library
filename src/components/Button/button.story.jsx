import React from 'react';
import { storiesOf } from '@storybook/react';
import centered from '@storybook/addon-centered/react';
import { ChevronDown, Play } from 'react-feather';
import Button from './index';

storiesOf('Button', module).addDecorator(centered)
	.add('Primary', () => (
		<div style={{ display: 'flex' }}>
			<Button primary>Primary</Button>
			<Button primary>
				<ChevronDown size="18" /> Primary
			</Button>
			<Button primary><Play size="18" /></Button>
			<Button primary disabled>Disabled</Button>
		</div>
	))
	.add('Secondary', () => (
		<div style={{ display: 'flex' }}>
			<Button>Secondary</Button>
			<Button><ChevronDown size="18" /> Secondary</Button>
			<Button><Play size="18" /></Button>
			<Button disabled>Disabled</Button>
		</div>
	));
