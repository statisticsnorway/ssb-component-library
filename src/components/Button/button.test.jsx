import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components'
import Button from './index';
import { ssbDark2, ssbGreen4, ssbWhite } from '../../style/colors';

describe('Button component', () => {
	test('Matches the snapshot', () => {
		const button = renderer.create(<Button text="Button" />).toJSON();
		expect(button).toMatchSnapshot();
	});

	test('Changes background if it is a primary button', () => {
		const button = renderer.create(<Button text="Button" primary />).toJSON();
		expect(button).toHaveStyleRule('background', ssbGreen4);
	});

	test('Primary button changes background if disabled', () => {
		const button = renderer.create(<Button text="Button" primary disabled />).toJSON();
		expect(button).toHaveStyleRule('background', ssbDark2);
		expect(button).toHaveStyleRule('color', ssbWhite);
	});

	test('Regular button changes background if disabled', () => {
		const button = renderer.create(<Button text="Button" disabled />).toJSON();
		expect(button).toHaveStyleRule('background', ssbWhite);
		expect(button).toHaveStyleRule('color', ssbDark2);
	});
});
