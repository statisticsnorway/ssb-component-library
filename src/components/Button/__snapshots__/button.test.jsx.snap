// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component Matches the snapshot 1`] = `
.c0 {
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #fff;
  border: border:1px solid #00824d;
  border-radius: 2px;
  color: #00824d;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  height: -webkit-fit-content;
  height: -moz-fit-content;
  height: fit-content;
  margin-right: 20px;
  padding: 10px 20px;
}

.c0:hover {
  background: #00824d;
  color: #fff;
}

.c0:active {
  background: #075745;
}

.c0:hover,
.c0:active {
  cursor: pointer;
}

.c0:disabled {
  cursor: not-allowed;
}

.c1 {
  font-family: 'Roboto',sans-serif;
  font-stretch: normal;
  font-family: 'Roboto',sans-serif;
  font-size: 16px;
  font-weight: bold;
  font-style: normal;
  font-stretch: normal;
  line-height: normal;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
  text-align: center;
}

<button
  className="c0 secondary"
  disabled={false}
  onClick={[Function]}
  type="button"
>
   
  <span
    className="c1 button-text"
  >
    Button
  </span>
</button>
`;
